# SQL <a id="head"></a>

Компания купила крупный сервис для чтения книг по подписке, предоставив информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг.

**Задача:** проанализировать базу данных, для формулировки ценностного предложения для нового продукта.

**Описание данных:**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

<h4>Содержание:<span class="tocSkip"></span></h4>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исследование-таблиц" data-toc-modified-id="Исследование-таблиц-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Исследование таблиц</a></span></li><li><span><a href="#Анализ-данных" data-toc-modified-id="Анализ-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Анализ данных</a></span><ul class="toc-item"><li><span><a href="#Количество-книг,-вышедших-после-1-января-2000-года" data-toc-modified-id="Количество-книг,-вышедших-после-1-января-2000-года-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Количество книг, вышедших после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзоров-и-средняя-оценка-книг" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-книг-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Количество обзоров и средняя оценка книг</a></span></li><li><span><a href="#Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книг</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
    'pwd': 'Sdf4$2;d-d30pp', # пароль
    'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
    'port': 6432, # порт подключения
    'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
     db_config['pwd'],
     db_config['host'],
     db_config['port'],
     db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Исследование таблиц 

In [2]:
sql='''
SELECT *
FROM books
LIMIT 5
'''
pd.io.sql.read_sql(sql, con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
sql='''
SELECT COUNT(*)
FROM books 
'''
pd.io.sql.read_sql(sql, con = engine)

count
0   1000

В таблице `books` хранится информация о 1000 книг.

In [4]:
sql='''
SELECT *
FROM authors
LIMIT 5
'''
pd.io.sql.read_sql(sql, con = engine) 

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
sql='''
SELECT COUNT(*)
FROM authors
'''
pd.io.sql.read_sql(sql, con = engine)

count
0    636

В таблице `authors` хранится информация о 636 авторах.

In [6]:
sql='''
SELECT *
FROM publishers
LIMIT 5
'''
pd.io.sql.read_sql(sql, con = engine) 

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [7]:
sql='''
SELECT COUNT(*)
FROM publishers
'''
pd.io.sql.read_sql(sql, con = engine)

count
0    340

В таблице `publishers` хранится информация о 340 издательств.

In [8]:
sql='''
SELECT *
FROM ratings
LIMIT 5
'''
pd.io.sql.read_sql(sql, con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [9]:
sql='''
SELECT COUNT(*)
FROM ratings
'''
pd.io.sql.read_sql(sql, con = engine)

count
0   6456

В таблице `ratings` хранится информация о 6456 пользовательских оценках книг.

In [10]:
sql='''
SELECT *
FROM reviews
LIMIT 5
'''
pd.io.sql.read_sql(sql, con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [11]:
sql='''
SELECT COUNT(*)
FROM reviews
'''
pd.io.sql.read_sql(sql, con = engine)

count
0   2793

В таблице `reviews` хранится информация о 2793 пользовательских обзорах на книги.

## Анализ данных

### Количество книг, вышедших после 1 января 2000 года

In [12]:
sql='''
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'
'''
pd.io.sql.read_sql(sql, con = engine) 

count
0    819

819 книг вышло после 1 января 2000 года.

### Количество обзоров и средняя оценка книг

In [13]:
sql='''
WITH rev_c as 
    (SELECT book_id, COUNT(book_id) as count_review
    FROM reviews
    GROUP BY book_id),
rat_a as
    (SELECT book_id, ROUND(AVG(rating), 2) as avg_rating
    FROM ratings
    GROUP BY book_id)

SELECT b.book_id, title, count_review, avg_rating
FROM books as b
JOIN rev_c on b.book_id=rev_c.book_id
JOIN rat_a on b.book_id=rat_a.book_id
ORDER BY avg_rating DESC
'''
pd.io.sql.read_sql(sql, con = engine) 

book_id                                              title  count_review  \
0         55          A Woman of Substance (Emma Harte Saga #1)             2   
1        518               Pop Goes the Weasel (Alex Cross  #5)             2   
2        347  In the Hand of the Goddess (Song of the Liones...             2   
3        513               Piercing the Darkness (Darkness  #2)             2   
4         57                    Act of Treason (Mitch Rapp  #9)             2   
..       ...                                                ...           ...   
989      915  The World Is Flat: A Brief History of the Twen...             3   
990      371                                              Junky             2   
991      316                  His Excellency: George Washington             2   
992      202                                      Drowning Ruth             3   
993      303                               Harvesting the Heart             2   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
989        2.25  
990        2.00  
991        2.00  
992        2.00  
993        1.50  

[994 rows x 4 columns]

Информация о 994 книгах. Максимальный рейтинг равен 5, минимальный 1.5

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [14]:
sql='''
WITH count_b as 
    (SELECT publisher_id, COUNT(book_id) AS count
    FROM books
    WHERE num_pages > 50
    GROUP BY publisher_id
    ORDER BY count DESC)

SELECT publisher, count 
FROM publishers AS p
JOIN count_b on p.publisher_id=count_b.publisher_id
LIMIT 1
'''
pd.io.sql.read_sql(sql, con = engine) 

publisher  count
0  Penguin Books     42

Исключив из анализа брошюры (книги толще 50 страниц), получили издательство Penguin Books, которое выпустило наибольшее число книг (42шт).

### Автор с самой высокой средней оценкой книг

In [15]:
sql='''
WITH rat_avg AS
    (SELECT book_id, ROUND(AVG(rating), 2) AS avg_rating
    FROM ratings
    GROUP BY book_id 
    HAVING COUNT(book_id) > 50),
auth AS
    (SELECT author_id, ROUND(AVG(avg_rating), 2) AS rating_avg
    FROM rat_avg
    JOIN books AS b ON rat_avg.book_id = b.book_id
    GROUP BY author_id) 
SELECT authors.author, rating_avg
FROM auth
JOIN authors ON authors.author_id = auth.author_id
ORDER BY rating_avg DESC
LIMIT 1
'''
pd.io.sql.read_sql(sql, con = engine)

author  rating_avg
0  J.K. Rowling/Mary GrandPré        4.29

Учитывая только книги с 50 и более оценками, автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré (4.29)

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [16]:
sql='''
WITH user_count AS
    (SELECT COUNT(username) AS count
    FROM reviews
    WHERE username in (SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(username) > 50)
    GROUP BY username)
SELECT ROUND(AVG(count))
FROM user_count
'''
pd.io.sql.read_sql(sql, con = engine) 

round
0   24.0

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок, равно 24.

## Вывод

Для компании, которая купила крупный сервис для чтения книг по подписке, была проанализирована база данных, для формулировки ценностного предложения для нового продукта.

Анализ показал, что:
* В базе содержится информация о 1000 книг, 636 авторах, 340 издательствах, 6456 пользовательских оценках и 2793  пользовательских обзорах на книг;
* 819 книг из 1000 были изданы после 1 января 2000;
* 42 книги выпустило издательство Penguin Books;
* Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré; 
* Среднее количество обзоров от пользователей, которые поставили больше 50 оценок, равно 24.

[Начало](#head)